pkg/spec/util.go

```diff
@@ -0,0 +1,214 @@
+// in a recursive manner.
+func getEmbeddedStructs(inputStruct interface{}) ([]interface{}, error) {
+
+	// Checking if input is pointer to struct
+	if err := isPointerToStruct(inputStruct); err != nil {
+		return nil, errors.Wrap(err, "Input parameter type mismatch")
+	}
+
+	var embeddedStructs []interface{}
+
+	val := reflect.ValueOf(inputStruct).Elem()
+
+	for i := 0; i < val.NumField(); i++ {
+		// Checking for embedded structs
+		if val.Type().Field(i).Anonymous {
+
+			// Appending pointer to the embedded/anonymous struct to
+			// embeddedStructs
+			embeddedStructs = append(embeddedStructs, val.Field(i).Addr().Interface())
+
+			// Since the current field is an anonymous struct, we call the
+			// function recursively to get embedded structs under it
+			recursiveEmbeddedStructs, err := getEmbeddedStructs(val.Field(i).Addr().Interface())
+			if err != nil {
+				return nil, errors.Wrapf(err, "Unable to get embedded structs recursively from %v", val.Field(i).String())
+			}
+
+			// Adding the returned recursive structs to embeddedStructs
+			embeddedStructs = append(embeddedStructs, recursiveEmbeddedStructs...)
+		}
+	}
+	return embeddedStructs, nil
+}
+
+// This function returns an array of all the JSON tags used for unmarshalling
+// (which are the first ones specified) which have a JSON tag "conflicting", in
+// all of the input structs.
+// It's made sure that only the tags in "package spec" are checked and all other
+// packages are ignored. This is done because we are expecting the structs to
+// be handled, i.e. marked as "conflicting" only in spec.go, and we do not want
+// to populate the list with JSON tags from some other package which had the
+// "conflicting" JSON tag for some other reason
+func getMarkedAsConflictingJSONUnmarshalTags(inputStructs []interface{}) ([]string, error) {
+	var blacklistedTags []string
+	for _, inputStruct := range inputStructs {
+
+		// Checking if input is pointer to struct
+		if err := isPointerToStruct(inputStruct); err != nil {
+			return nil, errors.Wrap(err, "Input parameter type mismatch")
+		}
+
+		val := reflect.ValueOf(inputStruct).Elem()
+
+		// Proceeding only if the struct belongs to the kedge's spec package
+		if val.Type().PkgPath() == "github.com/kedgeproject/kedge/pkg/spec" {
+			for i := 0; i < val.NumField(); i++ {
+				field := val.Type().Field(i)
+
+				fieldJSONTags := strings.Split(field.Tag.Get("json"), ",")
+
+				// Adding fields that have "conflicting" as a JSON tag to the
+				// array. We need not check the first tag that is used for
+				// unmarshalling.
+				for _, tag := range fieldJSONTags[1:] {
+					if tag == "conflicting" {
+						blacklistedTags = append(blacklistedTags, fieldJSONTags[0])
+					}
+				}
+			}
+		}
+	}
+	return blacklistedTags, nil
+}
+
+// This function checks if the input parameter has the pointer to a struct as
+// its underlying type.
+// Returns "nil" if true, or an error if false
+func isPointerToStruct(input interface{}) error {
```

Comment:
nit: ideally functions with `isFooBar` types should return `boolean`

because the error we are getting does not matter to us in a way we just wanna check if it is struct or not.

But also makes sense to return error because it helps us understand what are we passing!